### PR TITLE
Two-phase SQL rewrites

### PR DESCRIPTION
**Pull request by hrach**
It's kinda funny how this change looks simple. It took me 5 hours to come up with it
🤯💥🤯💥🤯💥🤯💥🤯💥🤯💥

### CODE REVIEW

File: tests/cases/integration/Relationships/relationships.oneHasMany.phpt [hunk 1]
```diff
@@ -342,12 +353,12 @@ class RelationshipOneHasManyTest extends DataTestCase
...
-			['tags->id' => [1]],
-			['tags->id' => null], // no match
+			['tags->id' => [1]], // #1
+			['tags->id' => null], // matches a book without tags (#4)
 		]);
 
-		Assert::same(1, $books->countStored());
-		Assert::same(1, $books->count());
+		Assert::same(2, $books->countStored());
+		Assert::same(2, $books->count());
```
Comment (hrach) [inline tests/cases/integration/Relationships/relationships.oneHasMany.phpt:361]: The changed behavior here actually matches Dbal behavior in Nextras 4. The original behavior matched Array behavior in Nextras 4 (i.e. the behavior was inconsistent). Are we ok with this changed behavior? 🤔 Most probably yes. cc @stepapo

File: src/Collection/Functions/BaseCompareFunction.php [hunk 2]
```diff
@@ -34,6 +34,13 @@ public function processArrayExpression(
 		}
 
 		if ($valueReference->aggregator !== null) {
+			if ($targetValue === null && $valueReference->value === [] && $this instanceof CompareEqualsFunction) {
```
Comment (hrach) [inline src/Collection/Functions/BaseCompareFunction.php:37]: Nasty fix, I have to realize how fix it better.

File: src/Collection/Functions/Result/DbalExpressionResult.php [hunk 3]
```diff
@@ -161,11 +166,22 @@ public function withHavingArgs(string $havingExpression, array $havingArgs): Dba
 	}
 
 
+	public function collect(ExpressionContext $context): DbalExpressionResult
```
Comment (hrach) [inline src/Collection/Functions/Result/DbalExpressionResult.php:169]: Design decision - make the DbalExpressionResult configurable with collectFun or override it in custom impl for and/or?

File: src/Collection/Functions/Result/DbalExpressionResult.php [hunk 4]
```diff
@@ -56,6 +59,8 @@ public function __construct(
 		public readonly ?PropertyMetadata $propertyMetadata = null,
 		?callable $valueNormalizer = null,
 		public readonly string|array|null $dbalModifier = null,
+		public readonly ?ExpressionContext $context = null,
```
Comment (hrach) [inline src/Collection/Functions/Result/DbalExpressionResult.php:62]: nit: added mistakenly, not needed; or is havignExpression === null a good detection of having expression? 🤔

File: src/Collection/Functions/Result/DbalExpressionResult.php [hunk 5]
```diff
@@ -3,6 +3,7 @@
 namespace Nextras\Orm\Collection\Functions\Result;
 
 
+use mysql_xdevapi\Expression;
```
Comment (hrach) [inline src/Collection/Functions/Result/DbalExpressionResult.php:6]: nit

File: tests/cases/integration/Relationships/relationships.oneHasMany.phpt (hunk 1 above)
Comment (stepapo) [reply tests/cases/integration/Relationships/relationships.oneHasMany.phpt:350]: Surely this is ok, makes things easier. Before this, `[CompareEqualsFunction::class, [CountAggregateFunction::class, 'tags->id'], 0]` had to be used for matching book with no tags.

File: src/Collection/Functions/BaseCompareFunction.php (hunk 2 above)
Comment (hrach) [reply src/Collection/Functions/BaseCompareFunction.php]: Reworked

**Pull request merged**